File: src/csharp/api/Repositories/IPatientRepository.cs
using Example.Api.Models;

namespace Example.Api.Repositories;

/// <summary>
/// Data repository interface for managing patients.
/// </summary>
public interface IPatientRepository
{
    /// <summary>
    /// Retrieves patients created within the specified time range with pagination.
    /// </summary>
    /// <param name="startTime">The start time of the range.</param>
    /// <param name="endTime">The end time of the range.</param>
    /// <param name="pageNumber">The page number to retrieve.</param>
    /// <param name="pageSize">The number of items per page.</param>
    /// <returns>A tuple containing the list of patients and the total count.</returns>
    Task<(IEnumerable<Patient> Data, long TotalCount)> GetPatientsAsync(
        DateTimeOffset startTime,
        DateTimeOffset endTime,
        int pageNumber,
        int pageSize);

    /// <summary>
    /// Checks if a patient exists by their id.
    /// </summary>
    /// <param name="id">The id of the patient.</param>
    /// <returns>True if the patient exists, otherwise false.</returns>
    Task<bool> IsExistPatientAsync(long id);

    /// <summary>
    /// Retrieves a patient by their id.
    /// </summary>
    /// <param name="id">The id of the patient.</param>
    /// <returns>The patient with the specified id, or null if not found.</returns>
    Task<Patient?> GetPatientAsync(long id);

    /// <summary>
    /// Retrieves a patient by their name.
    /// </summary>
    /// <param name="name">The name of the patient.</param>
    /// <returns>The patient with the specified name, or null if not found.</returns>
    Task<Patient?> GetPatientByNameAsync(string name);

    /// <summary>
    /// Creates a new patient record.
    /// </summary>
    /// <param name="patient">The patient entity to create.</param>
    /// <returns>The created patient entity.</returns>
    Task<Patient> AddAsync(Patient patient);
}

File: src/nodejs/src/repositories/IPatientRepository.ts
import { Patient } from "../entities/Patient";

/**
 * Interface for patient repository operations.
 */
export interface IPatientRepository {
  /**
   * Retrieves a paginated list of patients within a date range.
   * @param startTime - The start of the date range.
   * @param endTime - The end of the date range.
   * @param pageNumber - The page number to retrieve.
   * @param pageSize - The number of items per page.
   * @returns A tuple containing the patients array and total count.
   */
  getPatientsAsync(startTime: Date, endTime: Date, pageNumber: number, pageSize: number): Promise<[Patient[], number]>;

  /**
   * Checks if a patient exists by ID.
   * @param id - The patient ID.
   * @returns True if the patient exists, false otherwise.
   */
  isExistPatientAsync(id: number): Promise<boolean>;

  /**
   * Retrieves a patient by ID.
   * @param id - The patient ID.
   * @returns The patient if found, null otherwise.
   */
  getPatientAsync(id: number): Promise<Patient | null>;

  /**
   * Retrieves a patient by name.
   * @param name - The patient name.
   * @returns The patient if found, null otherwise.
   */
  getPatientByNameAsync(name: string): Promise<Patient | null>;

  /**
   * Adds a new patient to the repository.
   * @param patient - The patient to add.
   * @returns The added patient with generated ID.
   */
  addAsync(patient: Patient): Promise<Patient>;
}
